File: ios/Sources/NexdoCore/TodayActionQueue.swift
import Foundation

public struct TodayActionQueue: Sendable {
    public static let windowMinutes = 15
    public let primaryAction: TaskAction?
    public let nextActions: [TaskAction]
    public let laterActions: [TaskAction]
    public let overdueCount: Int
    public let deferredTaskIDs: Set<String>
    public var immediateTaskIDs: Set<String> { Set(([primaryAction].compactMap { $0 } + nextActions).map(\.taskId)) }
    public var reservedTaskIDs: Set<String> { immediateTaskIDs.union(laterActions.map(\.taskId)).union(deferredTaskIDs) }
    public var hasImmediateActions: Bool { primaryAction != nil || !nextActions.isEmpty }
    public var recommendation: String? {
        primaryAction.map { "You have a follow-up with \($0.contactName) due now. Contact them before starting your next task." }
    }

    public init(actions: [TaskAction], tasks: [NexdoTask], now: Date, timeZone: TimeZone,
                windowMinutes: Int = Self.windowMinutes) {
        var calendar = Calendar(identifier: .gregorian); calendar.timeZone = timeZone
        let end = calendar.date(byAdding: .day, value: 1, to: calendar.startOfDay(for: now))!
        let window = now.addingTimeInterval(Double(max(0, windowMinutes)) * 60)
        let active = tasks.filter { !$0.isDone && $0.status != "CANCELLED" }
        let priorities = Dictionary(active.map { ($0.id, Self.priority($0.priority)) }, uniquingKeysWith: max)
        deferredTaskIDs = Set(actions.filter {
            priorities[$0.taskId] != nil && $0.status != .completed && $0.status != .cancelled &&
            ($0.notificationDate ?? .distantPast) >= end
        }.map(\.taskId))
        var seen = Set<String>()
        let ordered = actions.filter {
            priorities[$0.taskId] != nil && $0.status != .completed && $0.status != .cancelled &&
            $0.notificationDate != nil && $0.notificationDate! < end
        }.sorted {
            if $0.notificationDate != $1.notificationDate { return $0.notificationDate! < $1.notificationDate! }
            if priorities[$0.taskId] != priorities[$1.taskId] { return priorities[$0.taskId, default: 0] > priorities[$1.taskId, default: 0] }
            return $0.id < $1.id
        }.filter { seen.insert($0.taskId).inserted }
        let due = ordered.filter { $0.notificationDate! <= now }
        primaryAction = due.first
        nextActions = ordered.filter { $0.id != due.first?.id && $0.notificationDate! <= window }
        laterActions = ordered.filter { $0.notificationDate! > window }
        overdueCount = due.filter { $0.notificationDate! < now }.count
    }
    private static func priority(_ value: String) -> Int {
        switch value { case "CRITICAL": 3; case "HIGH": 2; case "LOW": 0; default: 1 }
    }
}

File: mobile/src/lib/todayActionQueue.ts
import type { NexdoTask } from '../api/types';
import { addDays, isDone, startOfDay } from './taskQuery';

/**
 * Port of `TodayActionQueue` (ios/Sources/NexdoCore/TodayActionQueue.swift).
 *
 * This is a PURE function of the actions and the tasks, so it is ported and tested now even though
 * nothing supplies `actions` yet. `TaskActionCoordinator`, the only source, keeps its state in a local
 * JSON file and schedules local notifications through `UNUserNotificationCenter`, resolving people
 * through `Contacts` — all of which is Phase 8. Until then every list here is empty and
 * `hasImmediateActions` is false, which is exactly what the Today dashboard renders today.
 */

/** `TaskActionStatus` (ios/Sources/NexdoCore/TaskAction.swift:7-9). */
export type TaskActionStatus =
  | 'pending'
  | 'scheduled'
  | 'awaitingApproval'
  | 'approved'
  | 'executing'
  | 'completed'
  | 'failed'
  | 'cancelled';

/** `TaskActionChannel` (TaskAction.swift:6). */
export type TaskActionChannel = 'call' | 'message' | 'email';

/** `TaskAction` (TaskAction.swift:21-34), with the fields the queue reads. */
export type TaskAction = {
  id: string;
  taskId: string;
  contactName: string;
  preferredAction?: TaskActionChannel | null;
  status: TaskActionStatus;
  /** `scheduledAt`, epoch ms. */
  scheduledAt?: number | null;
  /** `snoozedUntil`, epoch ms. */
  snoozedUntil?: number | null;
  sourceTitle: string;
};

/** `notificationDate` (TaskAction.swift:34): a snooze wins over the original schedule. */
export function notificationDate(action: TaskAction): number | null {
  return action.snoozedUntil ?? action.scheduledAt ?? null;
}

export type TodayActionQueue = {
  primaryAction: TaskAction | null;
  nextActions: TaskAction[];
  laterActions: TaskAction[];
  overdueCount: number;
  deferredTaskIds: Set<string>;
  immediateTaskIds: Set<string>;
  reservedTaskIds: Set<string>;
  hasImmediateActions: boolean;
  /** `recommendation` (TodayActionQueue.swift:13-15). */
  recommendation: string | null;
};

/** `windowMinutes` (TodayActionQueue.swift:4). */
export const ACTION_WINDOW_MINUTES = 15;

/** `priority(_:)` (TodayActionQueue.swift:45-47). */
function priorityRank(value: string): number {
  switch (value) {
    case 'CRITICAL':
      return 3;
    case 'HIGH':
      return 2;
    case 'LOW':
      return 0;
    default:
      return 1;
  }
}

/** `TodayActionQueue.init` (TodayActionQueue.swift:17-44). */
export function buildActionQueue({
  actions,
  tasks,
  now,
  timeZone,
  windowMinutes = ACTION_WINDOW_MINUTES,
}: {
  actions: TaskAction[];
  tasks: NexdoTask[];
  now: number;
  timeZone: string;
  windowMinutes?: number;
}): TodayActionQueue {
  // `end` is the start of TOMORROW: the queue only ever covers the rest of today.
  const end = addDays(startOfDay(now, timeZone), 1, timeZone);
  const window = now + Math.max(0, windowMinutes) * 60_000;

  const active = tasks.filter((task) => !isDone(task) && task.status !== 'CANCELLED');
  // `uniquingKeysWith: max` — a duplicate id keeps the HIGHEST priority.
  const priorities = new Map<string, number>();
  for (const task of active) {
    const rank = priorityRank(task.priority);
    priorities.set(task.id, Math.max(priorities.get(task.id) ?? rank, rank));
  }

  const live = (action: TaskAction) =>
    priorities.has(action.taskId) && action.status !== 'completed' && action.status !== 'cancelled';

  const deferredTaskIds = new Set(
    actions.filter((action) => live(action) && (notificationDate(action) ?? Number.NEGATIVE_INFINITY) >= end).map((action) => action.taskId),
  );

  const seen = new Set<string>();
  const ordered = actions
    .filter((action) => {
      const at = notificationDate(action);
      return live(action) && at !== null && at < end;
    })
    .sort((left, right) => {
      const leftAt = notificationDate(left) as number;
      const rightAt = notificationDate(right) as number;
      if (leftAt !== rightAt) return leftAt - rightAt;
      const leftRank = priorities.get(left.taskId) ?? 0;
      const rightRank = priorities.get(right.taskId) ?? 0;
      if (leftRank !== rightRank) return rightRank - leftRank;
      return left.id < right.id ? -1 : left.id > right.id ? 1 : 0;
    })
    // One action per task: the first in the sorted order wins.
    .filter((action) => {
      if (seen.has(action.taskId)) return false;
      seen.add(action.taskId);
      return true;
    });

  const due = ordered.filter((action) => (notificationDate(action) as number) <= now);
  const primaryAction = due[0] ?? null;
  const nextActions = ordered.filter(
    (action) => action.id !== primaryAction?.id && (notificationDate(action) as number) <= window,
  );
  const laterActions = ordered.filter((action) => (notificationDate(action) as number) > window);
  const overdueCount = due.filter((action) => (notificationDate(action) as number) < now).length;

  const immediateTaskIds = new Set([...(primaryAction ? [primaryAction] : []), ...nextActions].map((action) => action.taskId));
  const reservedTaskIds = new Set([
    ...immediateTaskIds,
    ...laterActions.map((action) => action.taskId),
    ...deferredTaskIds,
  ]);

  return {
    primaryAction,
    nextActions,
    laterActions,
    overdueCount,
    deferredTaskIds,
    immediateTaskIds,
    reservedTaskIds,
    hasImmediateActions: primaryAction !== null || nextActions.length > 0,
    recommendation: primaryAction
      ? `You have a follow-up with ${primaryAction.contactName} due now. Contact them before starting your next task.`
      : null,
  };
}
